File: JS/Tree/1261. Find Elements in a Contaminated Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 */
var FindElements = function(root) {
    root.val = 0;
    const stack = [root];
    const values = [];
    while(stack.length) {
        const n = stack.pop();
        if (n) {
            values.push(n.val);
            if (n.left) {
                n.left.val = 2 * n.val + 1;
                stack.push(n.left);
            }
            if (n.right) {
                n.right.val = 2 * n.val + 2;
                stack.push(n.right);
            }
        }
    }
    this.values = values;
};

/** 
 * @param {number} target
 * @return {boolean}
 */
FindElements.prototype.find = function(target) {
    return this.values.indexOf(target) > -1;
};

/** 
 * Your FindElements object will be instantiated and called as such:
 * var obj = new FindElements(root)
 * var param_1 = obj.find(target)
 */

File: Golang/Tree/1261.Find-elements-in-a-contaminated-binary-tree.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type FindElements struct {
	root *TreeNode
}

func Constructor(root *TreeNode) FindElements {
	root.Val = 0
	stack := []*TreeNode{root}
	for len(stack) > 0 {
		node := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		if node.Left != nil {
			node.Left.Val = 2*node.Val + 1
			stack = append(stack, node.Left)
		}
		if node.Right != nil {
			node.Right.Val = 2*node.Val + 2
			stack = append(stack, node.Right)
		}
	}
	return FindElements{
		root: root,
	}
}

func (this *FindElements) Find(target int) bool {
	stack := []*TreeNode{this.root}
	for len(stack) > 0 {
		node := stack[len(stack)-1]
		if node.Val == target {
			return true
		}
		stack = stack[:len(stack)-1]
		if node.Left != nil {
			stack = append(stack, node.Left)
		}
		if node.Right != nil {
			stack = append(stack, node.Right)
		}
	}
	return false
}
